bundle/config/mutator/log_resource_references.go

```diff
@@ -0,0 +1,145 @@
+package mutator
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/cli/libs/dyn"
+	"github.com/databricks/cli/libs/dyn/dynvar"
+	"github.com/databricks/cli/libs/log"
+	"github.com/databricks/cli/libs/structaccess"
+)
+
+// Longest field name:
+// >>> len('hard_deletion_sync_min_interval_in_seconds') == 42
+const maxFieldLength = 42
+
+const maxMetricLength = 200
+
+// logResourceReferences scans resources for ${resources.*} references and logs them.
+func LogResourceReferences() bundle.Mutator {
+	return &logResourceReferences{}
+}
+
+type logResourceReferences struct{}
+
+func (m *logResourceReferences) Name() string {
+	return "LogResourceReferences"
+}
+
+func (m *logResourceReferences) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	err := b.Config.Mutate(func(root dyn.Value) (dyn.Value, error) {
+		pattern := dyn.NewPattern(dyn.Key("resources"))
+		used := map[string]struct{}{}
+		_, err := dyn.MapByPattern(root, pattern, func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+			_ = dyn.WalkReadOnly(v, func(path dyn.Path, val dyn.Value) error {
+				ref, ok := dynvar.NewRef(val)
+				if !ok {
+					return nil
+				}
+				for _, r := range ref.References() {
+					// Only track ${resources.*} references.
+					if !strings.HasPrefix(r, "resources.") {
+						continue
+					}
+
+					key := convertReferenceToMetric(ctx, b.Config, r)
+					if key != "" {
+						used[key] = struct{}{}
+					}
+				}
+				return nil
+			})
+			return v, nil
+		})
+		if err != nil {
+			return dyn.InvalidValue, err
+		}
+
+		maxRefsLogged := 50
+
+		// map iteration is randomized, which works for this case
+		for key := range used {
+			b.Metrics.AddBoolValue(key, true)
+			maxRefsLogged--
+			if maxRefsLogged <= 0 {
+				break
+			}
+		}
+
+		return root, nil
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+// convertReferenceToMetric converts a reference like "resources.jobs.foo.id" to
+// a telemetry key like "resref__jobs__id"
+func convertReferenceToMetric(ctx context.Context, cfg any, ref string) string {
+	p, err := dyn.NewPathFromString(ref)
+	if err != nil || len(p) < 3 || p[0].Key() != "resources" {
+		return ""
+	}
+
+	kept := []string{"resref", truncate(p[1].Key(), maxFieldLength, "")}
+	for i := 3; i < len(p); i++ {
+		c := p[i]
+		if c.Key() != "" {
+			item := c.Key()
+
+			repl, err := censorValue(ctx, cfg, p[:i])
+			if err != nil {
+				kept[0] = "resreferr"
+				break
+			}
+
+			if repl != "" {
+				item = repl
+			}
+
+			item = truncate(item, maxFieldLength, "")
+			kept = append(kept, item)
+			continue
+		}
+	}
+
+	result := strings.Join(kept, "__")
+	return truncate(result, maxMetricLength, "__cut")
+}
+
+func truncate(s string, n int, suffix string) string {
```

Comment:
Just FYI: This would not work as intended if any field names have non-ASCII characters, since indexing a string directly in Go gives you the raw byte rather than the UTF-8 character. This is not a concern here since field names should not contain non-ASCII characters.